map/framework.cpp

```diff
@@ -3313,3 +3316,73 @@ void Framework::OnPowerSchemeChanged(power_management::Scheme const actualScheme
   if (actualScheme == power_management::Scheme::EconomyMaximum && GetTrafficManager().IsEnabled())
     GetTrafficManager().SetEnabled(false);
 }
+
+bool Framework::ShouldShowProducts() const
+{
+  auto const connectionStatus = Platform::ConnectionStatus();
+  if (connectionStatus == Platform::EConnectionType::CONNECTION_NONE)
+    return false;
+
+  std::string donateUrl;
+  if (!settings::Get(settings::kDonateUrl, donateUrl)) // donation is disabled
+    return false;
+
+  if (!m_usageStats.IsLoyalUser())
+    return false;
+
+  if (!storage::IsPointCoveredByDownloadedMaps(GetCurrentPlacePageInfo().GetMercator(), m_storage, *m_infoGetter))
+    return false;
+
+  #ifdef DEBUG
+  uint32_t constexpr kPopupCloseTimeout = 10;
+  uint32_t constexpr kProductSelectTimeout = 20;
+  #else
+  uint32_t constexpr kPopupCloseTimeout = 60 * 60 * 24  * 14; // 14 days
+  uint32_t constexpr kProductSelectTimeout = 180 * 24 * 60 * 60; // 180 days
+  #endif
+
+  uint64_t popupCloseTime;
+  if (!settings::Get(kPlacePageProductsPopupCloseTime, popupCloseTime))
+    popupCloseTime = 0; // popup was never shown
+  uint64_t productSelectTime;
+  if (!settings::Get(kPlacePageProductSelectionTime, productSelectTime))
+    productSelectTime = 0; // product was never selected
+
+  auto const now = base::SecondsSinceEpoch();
+  bool const closePopupTimeoutExpired = popupCloseTime + kPopupCloseTimeout < now;
+  bool const productSelectTimeoutExpired = productSelectTime + kProductSelectTimeout < now;
+  if (closePopupTimeoutExpired && productSelectTimeoutExpired)
+    return true;
+
+  return false;
+}
+
+std::optional<products::ProductsConfig> Framework::GetProductsConfiguration() const
+{
+  if (!ShouldShowProducts())
+    return nullopt;
+  return products::GetProductsConfiguration();
+}
+
+void Framework::DidCloseProductsPopup(ProductsPopupCloseReason reason) const
+{
+  auto const now = base::SecondsSinceEpoch();
+  settings::Set(kPlacePageProductsPopupCloseTime, now);
+  switch (reason)
+  {
+    case ProductsPopupCloseReason::Close:
+    case ProductsPopupCloseReason::RemindLater: // same behaviour as close
```

Comment:
The "Remind Later" and close "x" behavior should be different.
Now both lead to a 2 weeks delay till the next popup.

IMO its too long for "Remind Later", maybe a week?

People will choose the "x" button to close the popup if their reason is different from other available options (donate, remind later, already donated): don't want to donate, no money, etc.

There is no point to display the popup again soon to users who e.g. doesn't want to donate, but it could lead to unwanted irritation.

So its better to change the "x" close option to a long delay like it was in the initial version - maybe same as Already Donated 180 days.